File: robot_state/state_node.py
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Pose2D
from scipy.spatial.transform import Rotation as R


class StateNode(Node):

    def __init__(self):
        super().__init__("state_node")

        self.declare_parameter("odom_topic", "/odom1")
        self.declare_parameter("pose_topic", "/robot_pose")

        odom_topic = self.get_parameter("odom_topic").value
        pose_topic = self.get_parameter("pose_topic").value

        self.pose_pub = self.create_publisher(Pose2D, pose_topic, 10)
        self.create_subscription(Odometry, odom_topic, self.odom_callback, 10)

    def odom_callback(self, msg: Odometry) -> None:
        pose = Pose2D()
        pose.x = msg.pose.pose.position.x
        pose.y = msg.pose.pose.position.y

        q = msg.pose.pose.orientation
        yaw = R.from_quat([q.x, q.y, q.z, q.w]).as_euler("xyz")[2]
        pose.theta = yaw

        self.pose_pub.publish(pose)


def main():
    rclpy.init()
    node = StateNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/state_node.cpp
#include "robot_state/state_node.hpp"

#include "tf2/utils.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"

namespace robot_state {

StateNode::StateNode(const rclcpp::NodeOptions &options)
    : Node("state_node", options) {
  declare_parameter("odom_topic", "/odom1");
  declare_parameter("pose_topic", "/robot_pose");

  const auto odom_topic = get_parameter("odom_topic").as_string();
  const auto pose_topic = get_parameter("pose_topic").as_string();

  odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      odom_topic, 10,
      [this](const nav_msgs::msg::Odometry &msg) { onOdometry(msg); });

  pose_pub_ = create_publisher<geometry_msgs::msg::Pose2D>(pose_topic, 10);

  RCLCPP_INFO(get_logger(), "StateNode ready | %s → %s", odom_topic.c_str(),
              pose_topic.c_str());
}

void StateNode::onOdometry(const nav_msgs::msg::Odometry &msg) {
  geometry_msgs::msg::Pose2D pose;
  pose.x = msg.pose.pose.position.x;
  pose.y = msg.pose.pose.position.y;
  pose.theta = tf2::getYaw(msg.pose.pose.orientation);

  pose_pub_->publish(pose);
}

} // namespace robot_state
